pkg/comp-functions/runtime/function_mgr.go

```diff
@@ -1064,3 +1069,45 @@ func (s *ServiceRuntime) addUsages() error {
 	}
 	return nil
 }
+
+func (s *ServiceRuntime) ForwardEvents() error {
+	claimRef := s.observedComposite.GetClaimReference()
+	// Claim is not yet populated, retry next time
+	if claimRef == nil {
+		return nil
+	}
+	eventForwardValue := fmt.Sprintf("%s/%s/%s/%s", claimRef.APIVersion, claimRef.Kind, claimRef.Namespace, claimRef.Name)
+	for _, res := range s.desiredResources {
+		r := res.Resource
+
+		// For kube objects set the annotation 'EventForwardAnnotation' in managed resource and not in kube object itself
+		if isKubeObject(r) {
+			p := "spec.forProvider.manifest.metadata.annotations"
+			v, _ := r.GetValue(p)
+			mrAnnotations := make(map[string]any)
+			if v != nil {
+				mrAnnotations = v.(map[string]any)
+			}
+			mrAnnotations[EventForwardAnnotation] = eventForwardValue
+			err := r.SetValue(p, mrAnnotations)
+			if err != nil {
+				return fmt.Errorf("cannot set event forward annotations for managed object %s: %w", r.GetName(), err)
+			}
+
+			// For non kube objects set directly into the managed resource
+		} else {
```

Comment:
Shouldn't we also add the annotation to the outer kubeobject as well? Sometimes they might also fail to actually apply an object due to webhooks or other issues.

Reply:
wouldn't that cause doubled entries in 99% of cases? outer object contains inner object's events and that would make readibility much worse

Reply:
Not necessarily, the outer object is more or less only concerned with applying the inner object to the cluster. For example if there's a webhook that blocks `.spec.important = false`, or if you try to apply a namespaced object without a namespace. That would only show up in the outer object. In such a case the object would not even reach the kubernetes api, and we'd have to describe the kubeobject to figure out why.

The events of the inner object are created by the controller that handles said object, only after its creation.

Reply:
I will add them to the kube Objects as well